FrameMetadata/app/src/main/java/com/tokbox/android/tutorials/framemetadata/MainActivity.java

```diff
@@ -0,0 +1,310 @@
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+
+        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this);
+    }
+
+    @Override
+    public void onPermissionsGranted(int requestCode, List<String> perms) {
+        Log.d(TAG, "onPermissionsGranted:" + requestCode + ":" + perms.size());
+    }
+
+    @Override
+    public void onPermissionsDenied(int requestCode, List<String> perms) {
+        Log.d(TAG, "onPermissionsDenied:" + requestCode + ":" + perms.size());
+
+        if (EasyPermissions.somePermissionPermanentlyDenied(this, perms)) {
+            new AppSettingsDialog.Builder(this)
+                    .setTitle(getString(R.string.title_settings_dialog))
+                    .setRationale(getString(R.string.rationale_ask_again))
+                    .setPositiveButton(getString(R.string.setting))
+                    .setNegativeButton(getString(R.string.cancel))
+                    .setRequestCode(RC_SETTINGS_SCREEN_PERM)
+                    .build()
+                    .show();
+        }
+    }
+
+    @AfterPermissionGranted(RC_VIDEO_APP_PERM)
+    private void requestPermissions() {
+        String[] perms = { Manifest.permission.INTERNET, Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO };
+        if (EasyPermissions.hasPermissions(this, perms)) {
+            mSession = new Session.Builder(MainActivity.this, OpenTokConfig.API_KEY, OpenTokConfig.SESSION_ID).build();
+            mSession.setSessionListener(this);
+            mSession.connect(OpenTokConfig.TOKEN);
+        } else {
+            EasyPermissions.requestPermissions(this, getString(R.string.rationale_video_app), RC_VIDEO_APP_PERM, perms);
+        }
+    }
+
+    /**
+     * @return yyyy-MM-dd'T'HH:mm:ssZZZZZ
+     */
+    private static String getCurrentTimeStamp(){
+        try {
+
+            SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZZZZZ");
+            String currentDateTime = dateFormat.format(new Date());
+
+            return currentDateTime;
+        } catch (Exception e) {
+            e.printStackTrace();
+
+            return null;
+        }
+    }
+
+    @Override
+    public void onConnected(Session session) {
+        Log.d(TAG, "onConnected: Connected to session " + session.getSessionId());
+
+        // capturer
+        CustomVideoCapturer capturer = new CustomVideoCapturer(MainActivity.this, Publisher.CameraCaptureResolution.MEDIUM, Publisher.CameraCaptureFrameRate.FPS_30);
+        capturer.setCustomVideoCapturerDataSource(new CustomVideoCapturer.CustomVideoCapturerDataSource() {
+            @Override
+            public byte[] retrieveMetadata() {
+                return getCurrentTimeStamp().getBytes();
+            }
+        });
+
+        // renderer
+        InvertedColorsVideoRenderer renderer = new InvertedColorsVideoRenderer(MainActivity.this);
```

Comment:
this is the changing piece.

Reply:
I know that our current custom renderer is `InvertedColorsVideoRenderer`... but it feels a lot strange to use it here.

wdyt?

I'm thiking on changing the Custom-Video-Driver sample to have a CustomRenderer and this Inverted too. This sample will use the CustomRenderer (without inversion).